src/shims/sync.rs

```diff
@@ -392,41 +405,39 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                 {
                     this.eval_libc_i32("EBUSY")
                 } else if kind == this.eval_libc("PTHREAD_MUTEX_RECURSIVE")? {
-                    match locked_count.checked_add(1) {
-                        Some(new_count) => {
-                            mutex_set_locked_count(this, mutex_op, Scalar::from_u32(new_count))?;
-                            Ok(0)
-                        }
-                        None => this.eval_libc_i32("EAGAIN"),
-                    }
+                    this.mutex_lock(id, active_thread);
+                    Ok(0)
                 } else {
                     throw_ub_format!(
                         "called pthread_mutex_trylock on an unsupported type of mutex"
                     );
                 }
             }
+        } else {
+            // The mutex is unlocked. Let's lock it.
+            this.mutex_lock(id, active_thread);
+            Ok(0)
         }
     }
 
     fn pthread_mutex_unlock(&mut self, mutex_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
         let this = self.eval_context_mut();
 
         let kind = mutex_get_kind(this, mutex_op)?.not_undef()?;
-        let locked_count = mutex_get_locked_count(this, mutex_op)?.to_u32()?;
-        let owner_thread: ThreadId = mutex_get_owner(this, mutex_op)?.to_u32()?.into();
-
-        if owner_thread != this.get_active_thread()? {
-            throw_ub_format!("called pthread_mutex_unlock on a mutex owned by another thread");
-        } else if locked_count == 1 {
-            let blockset = mutex_get_or_create_blockset(this, mutex_op)?;
-            if let Some(new_owner) = this.unblock_some_thread(blockset)? {
-                // We have at least one thread waiting on this mutex. Transfer
-                // ownership to it.
-                mutex_set_owner(this, mutex_op, new_owner.to_u32_scalar())?;
-            } else {
-                // No thread is waiting on this mutex.
-                mutex_set_owner(this, mutex_op, Scalar::from_u32(0))?;
-                mutex_set_locked_count(this, mutex_op, Scalar::from_u32(0))?;
+        let id = mutex_get_or_create_id(this, mutex_op)?;
+
+        if let Some((owner_thread, current_locked_count)) = this.mutex_unlock(id) {
+            if owner_thread != this.get_active_thread()? {
+                throw_ub_format!("called pthread_mutex_unlock on a mutex owned by another thread");
+            }
+            if current_locked_count == 0 {
+                // The mutex is unlocked.
+                if let Some(thread) = this.mutex_dequeue(id) {
+                    // We have at least one thread waiting on this mutex. Transfer
+                    // ownership to it.
+                    this.mutex_lock(id, thread);
+                    this.unblock_thread(thread)?;
```

Comment:
Every caller of `mutex_unlock` will need to repeat this logic, right? Wouldn't it be better to have that logic in `mutex_unlock` so that it is shared (also with an eventual Windows implementation of these)?